thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/diffsummary/MultiDimensionalSummary.java

```diff
@@ -0,0 +1,297 @@
+  private static final ThirdEyeCacheRegistry CACHE_REGISTRY_INSTANCE = ThirdEyeCacheRegistry.getInstance();
+
+  public static final String TOP_K_POSTFIX = "_topk";
+
+  private OLAPDataBaseClient olapClient;
+  private CostFunction costFunction;
+  private DateTimeZone dateTimeZone;
+
+  public MultiDimensionalSummary(OLAPDataBaseClient olapClient, CostFunction costFunction,
+      DateTimeZone dateTimeZone) {
+    Preconditions.checkNotNull(olapClient);
+    Preconditions.checkNotNull(dateTimeZone);
+    Preconditions.checkNotNull(costFunction);
+    this.olapClient = olapClient;
+    this.costFunction = costFunction;
+    this.dateTimeZone = dateTimeZone;
+  }
+
+  public SummaryResponse buildSummary(String dataset, String metric, long currentStartInclusive,
+      long currentEndExclusive, long baselineStartInclusive, long baselineEndExclusive, Dimensions dimensions,
+      Multimap<String, String> dataFilters, int summarySize, int depth, List<List<String>> hierarchies,
+      boolean doOneSideError) throws Exception {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(dataset));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(metric));
+    Preconditions.checkArgument(currentStartInclusive < currentEndExclusive);
+    Preconditions.checkArgument(baselineStartInclusive < baselineEndExclusive);
+    Preconditions.checkNotNull(dimensions);
+    Preconditions.checkArgument(dimensions.size() > 0);
+    Preconditions.checkNotNull(dataFilters);
+    Preconditions.checkArgument(summarySize > 1);
+    Preconditions.checkNotNull(hierarchies);
+    Preconditions.checkArgument(depth >= 0);
+
+    olapClient.setCollection(dataset);
+    List<MetricExpression> metricExpressions = Utils.convertToMetricExpressions(metric, MetricAggFunction.SUM, dataset);
+    olapClient.setMetricExpression(metricExpressions.get(0));
+    olapClient.setCurrentStartInclusive(new DateTime(currentStartInclusive, dateTimeZone));
+    olapClient.setCurrentEndExclusive(new DateTime(currentEndExclusive, dateTimeZone));
+    olapClient.setBaselineStartInclusive(new DateTime(baselineStartInclusive, dateTimeZone));
+    olapClient.setBaselineEndExclusive(new DateTime(baselineEndExclusive, dateTimeZone));
+
+    Cube cube = new Cube(costFunction);
+    SummaryResponse response;
+    if (depth > 0) { // depth != 0 means manual dimension order
+      cube.buildWithAutoDimensionOrder(olapClient, dimensions, dataFilters, depth, hierarchies);
+      Summary summary = new Summary(cube, costFunction);
+      response = summary.computeSummary(summarySize, doOneSideError, depth);
+    } else { // manual dimension order
+      cube.buildWithManualDimensionOrder(olapClient, dimensions, dataFilters);
+      Summary summary = new Summary(cube, costFunction);
+      response = summary.computeSummary(summarySize, doOneSideError);
+    }
+    response.setDataset(dataset);
+    response.setMetricName(metric);
+
+    return response;
+  }
+
+  private static Options buildOptions() {
```

Comment:
move this to a separate class, call it ContributionAnalysisTool

Reply:
Done!